zipnumcluster-ccpyspark.py

```diff
@@ -0,0 +1,263 @@
+        parser.add_argument("--num_lines", type=int, required=False,
+                            default=3000,
+                            help="number of lines to compress in each chunk")
+        parser.add_argument("--num_output_partitions", type=int, required=False,
+                            default=300,
+                            help="number of partitions/shards")
+
+    def run_job(self, session):
+        os.makedirs(self.args.output_base_url, exist_ok=True)
+        input = self.args.input_base_url + self.args.input
+        num_partitions = self.args.num_output_partitions
+        boundries_file_uri = self.args.partition_boundries_file
+        num_lines = self.args.num_lines
+        output_base_url = self.args.output_base_url
+        rdd = session.sparkContext.textFile(input).map(parse_line).filter(lambda x: x is not None)
+
+        # Cache the RDD with MEMORY_AND_DISK storage level
+        #rdd = rdd.persist(StorageLevel.MEMORY_AND_DISK)
+        #rdd = rdd.cache()
+
+        boundaries = None
+        ##logging.info(f"Boundaries file: {boundries_file_uri}")
+        if boundries_file_uri and self.check_for_output_file(boundries_file_uri):
+            ##logging.info(f"Boundaries file found, using it: {boundries_file_uri}")
+            with self.fetch_file(boundries_file_uri) as f:
+                boundaries = json.load(f)
+        else:
+            # this percent needs to be pretty small, since this collect brings data back to driver...
+            # 1/2 percent should be fine
+            samples = rdd.keys().sample(False, 0.005).collect()
+            samples.sort()
+            
+            # Ensure more even distribution by using quantiles
+            total_samples = len(samples)
+            boundaries = []
+            for i in range(1, num_partitions):
+                idx = (i * total_samples) // num_partitions
+                if idx < len(samples):
+                    boundaries.append(samples[idx])
+            
+            temp_file_name = 'temp_range_boundaries.json'
+            with open(temp_file_name, 'w') as f:
+                json.dump(boundaries, f)
+            
+            with open(temp_file_name, 'rb') as f:
+                self.write_output_file(boundries_file_uri, f)
+
+            os.unlink(temp_file_name)
+        
+        rdd = rdd.repartitionAndSortWithinPartitions(
+            numPartitions=num_partitions,
+            partitionFunc=lambda k: get_partition_id(k,boundaries),
+            keyfunc=lambda x: x[0]) \
+        .mapPartitionsWithIndex(lambda idx, iter: process_partition(idx, iter, num_lines, output_base_url)) \
+        .collect()
+    
+        # loop over the output files and concatenate them into a single final file
+        with open('cluster.idx', 'wb') as f:
+            for idx_file,_ in rdd:
+                with self.fetch_file(output_base_url + idx_file) as idx_fd:
```

Comment:
The default of `--output_base_url` is `my_cdx_bucket` (no trailing slash). So, when testing locally, I get the error:
```
4/12/20 14:03:37 INFO ZipNumClusterCdx: Reading local file my_cdx_bucketidx-00000.idx
Traceback (most recent call last):
  File ".../webarchive-indexing/zipnumcluster-ccpyspark.py", line 263, in <module>
    job.run()
  File ".../webarchive-indexing/sparkcc.py", line 187, in run
    self.run_job(session)
  File ".../webarchive-indexing/zipnumcluster-ccpyspark.py", line 252, in run_job
    with self.fetch_file(output_base_url + idx_file) as idx_fd:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ".../webarchive-indexing/sparkcc.py", line 727, in fetch_file
    warctemp = open(uri, 'rb')
               ^^^^^^^^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory: '.../webarchive-indexing/my_cdx_bucketidx-00000.idx'
```

Looks like, for testing locally, `--output_base_url` needs to be of the form: `file:/absolute/path/`.

Reply:
yes, that's correct (and is shown in the integration tests/scripts in the cc-redact project)